File: backend/internal/domain/repositories/product_repository.go
package repositories

import "catalogo-produtos/backend/internal/domain/entities"

// ProductRepository define as operações de persistência para produtos
type ProductRepository interface {
	Create(product *entities.Product) error
	GetByID(id uint) (*entities.Product, error)
	GetAll(filters *ProductFilter) ([]entities.Product, error)
	Update(product *entities.Product) error
	Delete(id uint) error
}

// ProductFilter define os filtros para busca de produtos
type ProductFilter struct {
	Name     string
	Category string
}

File: frontend/src/domain/repositories/ProductRepository.ts
import { Product } from '../entities/Product';

export interface ProductFilter {
  name?: string;
  category?: string;
}

// Interface do repositório de produtos
export interface ProductRepository {
  getProducts(filters?: ProductFilter): Promise<Product[]>;
  getProduct(id: number): Promise<Product>;
  createProduct(product: Omit<Product, 'id'>): Promise<Product>;
  updateProduct(id: number, product: Partial<Product>): Promise<Product>;
  deleteProduct(id: number): Promise<void>;
} 